sphinx/source/conf.py

```diff
@@ -201,3 +208,43 @@
 
 # If true, `todo` and `todoList` produce output, else they produce nothing.
 todo_include_todos = True
+
+
+# -- Docstring Improvements --------------------------------------------------
+
+
+# Regex code for typing replacements.
+# The "(?<![\.])" part checks to see if the string
+# starts with a period, and "\b" denotes word boundaries.
+# Only words that don't start with a period are replaced.
+_rt = [r"(?<![\.])(\b", r"\b)"]
+
+
+def autodoc_process_docstring(
+    app, what: str, name: str, obj, options, lines: List[str]
+) -> None:
+    """
+    Modify docstrings before creating html files.
+    Sphinx converts the 'Args:' and 'Returns:' sections of docstrings into
+    reStructuredText (rST) syntax, which can then be found via ':type' & ':rtype'.
+
+    See here for more information:
+    https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html
+    """
+    for i in range(len(lines)):
+        # Skip unless line is an parameter doc or a return doc
+        if not lines[i].startswith(":type"):
+            continue
+        if ":py:data:" in lines[i]:
+            continue
+
+        # Ensure Any, Callable, & Iterator types are hyperlinked with intersphinx.
+        # The tilde '~' character hides the 'typing.' portion of the string.
+        lines[i] = re.sub(_rt[0] + r"Any" + _rt[1], "~typing.Any", lines[i])
+        lines[i] = re.sub(_rt[0] + r"Callable" + _rt[1], "~typing.Callable", lines[i])
+        lines[i] = re.sub(_rt[0] + r"Iterator" + _rt[1], "~typing.Iterator", lines[i])
+        lines[i] = re.sub(_rt[0] + r"Iterable" + _rt[1], "~typing.Iterable", lines[i])
```

Comment:
why do we need to add `~typing.` for them? what is the exact situation that they cannot be identified?

Reply:
@aobo-y Adding `~typing.` to them ensures that intersphinx recognizes them properly, as it does not autodetect anything under the `typing` module. This way it shows up as hyperlinked (the `~typing.` portion doesn't show up int the HTML docs).

Reply:
hmm, then wondering if you have any clues why `Tensor` can be correctly linked with `from torch import Tensor` but not `Iterable` with `from typing import Iterable`?

Reply:
@aobo-y I'm not sure. Even inferring `Tensor` from the imports doesn't seem to be very robust as it only does in 3 types for the entire project.

These are the type hints before `autodoc_process_docstring`, with each line showing the name of the class / function and then the docstring type (`name`, `lines[i]`): https://gist.github.com/ProGamerGov/1d0636cd173f46effc8db7a2705f509d

And these are the type hints after: https://gist.github.com/ProGamerGov/31871f2d231ab68179ab18eecabb77b6

Reply:
Interesting. `TracInCPFast` and `TracInCPFastRandProj` are under the same file with the same docstring, but one is detected and one is not. Indeed pretty weird
```
captum.influence.TracInCPFast.influence :type targets: Tensor, optional
...
captum.influence.TracInCPFastRandProj.influence :type targets: :py:class:`~torch.Tensor`
```